Source/santactl/Commands/sync/SNTCommandSyncManager.m

```diff
@@ -252,12 +249,33 @@ - (void)processFCMMessage:(NSDictionary *)FCMmessage withMachineID:(NSString *)m
   }
 }
 
-- (NSData *)extractMessageDataFrom:(NSDictionary *)FCMmessage {
+- (NSData *)messageDataFromFCMmessage:(NSDictionary *)FCMmessage {
   if (![FCMmessage[@"data"] isKindOfClass:[NSDictionary class]]) return nil;
   if (![FCMmessage[@"data"][@"blob"] isKindOfClass:[NSString class]]) return nil;
   return [FCMmessage[@"data"][@"blob"] dataUsingEncoding:NSUTF8StringEncoding];
 }
 
+- (NSDictionary *)messageFromMessageData:(NSData *)messageData {
+  NSError *error;
+  NSDictionary *rawMessage = [NSJSONSerialization JSONObjectWithData:messageData
+                                                          options:0
+                                                            error:&error];
+  if (!rawMessage) {
+    LOGD(@"Unable to parse push notification message data: %@", error);
+    return nil;
+  }
+
+  // Create a new message dropping unexpected values
+  NSArray *allowedKeys = @[ kFCMActionKey, kFCMFileHashKey, kFCMFileNameKey, kFCMTargetHostIDKey ];
+  NSMutableDictionary *message = [NSMutableDictionary dictionaryWithCapacity:allowedKeys.count];
+  for (NSString *key in allowedKeys) {
+    if ([rawMessage[key] isKindOfClass:[NSString class]] && [rawMessage[key] length]) {
+      message[key] = rawMessage[key];
+    }
+  }
+  return message.count ? [message copy] : nil;
```

Comment:
Wouldn't returning nil here also count as being unable to parse, or is that expected in some cases?